osp/views/question.py

```diff
@@ -0,0 +1,73 @@
+from rest_framework import viewsets, status
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+
+from osp.models import (
+    Form, Question, Choice,
+    Checkbox, Dropdown, Paragraph, ShortAnswer,
+    Date, Time, FileUpload
+)
+from osp.serializers.question import QuestionSerializer
+from osp.serializers.fields import (
+    ChoiceSerializer, CheckboxSerializer, DropdownSerializer,
+    ParagraphSerializer, ShortAnswerSerializer, ParagraphSerializer,
+    DateSerializer, TimeSerializer, FileUploadSerializer
+)
+from osp.permissions.admin import IsAdmin
+from osp.utils.question import get_model_and_serializer
+
+class QuestionView(viewsets.ViewSet):
+
+    serializer_class = QuestionSerializer
+
+    # override permissions for different request methods
+    def get_permissions(self):
+        if self.request.method == 'GET':
+            self.permission_classes = [IsAuthenticated]
+        else:
+            self.permission_classes = [IsAuthenticated, IsAdmin]
+
+        return super(QuestionView, self).get_permissions()
+
+    
+    # GET request
+    def list(self, request):
+        form_id = self.request.query_params.get('form_id', None)
+        fields = Form.objects.filter(id=form_id).values_list('form_fields', flat=True)
+        objs = Question.objects.filter(id__in=fields).order_by('order')
+        results = []
+
+        for obj in objs:
+            value = get_model_and_serializer(obj.data_type)
+            model = value['model']
+            serializer = value['serializer']
+            instance = model.objects.get(id=obj.id)
+            results.append(serializer(instance).data)
+            
+        #response
+        return Response(results)
+
+
+    # POST request
+    def create(self, request):
+        form_id = self.request.query_params.get('form_id', None)
```

Comment:
query params are used for sorting/fetching purposes, if you need to create/update data, it needs to be sent in the body

Reply:
Also if you send it inside body of each question obj then I think line 69 `form.form_fields.add()` would also become redundant.